src/components/application_manager/rpc_plugins/vehicle_info_plugin/src/commands/mobile/subscribe_vehicle_data_request.cc

```diff
@@ -201,20 +201,39 @@ bool SubscribeVehicleDataRequest::CheckSubscriptionStatus(
 }
 
 bool SubscribeVehicleDataRequest::SubscribePendingVehicleData(
-    ApplicationSharedPtr app, const smart_objects::SmartObject& msg_params) {
+    ApplicationSharedPtr app, smart_objects::SmartObject& msg_params) {
   LOG4CXX_DEBUG(logger_, "Subscribing to all pending VehicleData");
 
-  for (const auto& vi_name : vi_waiting_for_subscribe_) {
+  std::set<hmi_apis::Common_VehicleDataResultCode::eType> skiped_result_codes(
+      {hmi_apis::Common_VehicleDataResultCode::VDRC_TRUNCATED_DATA,
+       hmi_apis::Common_VehicleDataResultCode::VDRC_DISALLOWED,
+       hmi_apis::Common_VehicleDataResultCode::VDRC_USER_DISALLOWED,
+       hmi_apis::Common_VehicleDataResultCode::VDRC_INVALID_ID,
+       hmi_apis::Common_VehicleDataResultCode::VDRC_DATA_NOT_AVAILABLE,
+       hmi_apis::Common_VehicleDataResultCode::VDRC_DATA_NOT_SUBSCRIBED,
+       hmi_apis::Common_VehicleDataResultCode::VDRC_IGNORED});
+
+  for (auto vi_name = vi_waiting_for_subscribe_.begin();
+       vi_name != vi_waiting_for_subscribe_.end();) {
     const bool is_subscription_successful = CheckSubscriptionStatus(
-        ConvertRequestToResponseName(vi_name), msg_params);
+        ConvertRequestToResponseName(*vi_name), msg_params);
 
     if (is_subscription_successful) {
       auto& ext = VehicleInfoAppExtension::ExtractVIExtension(*app);
-      ext.subscribeToVehicleInfo(vi_name);
-      vi_waiting_for_subscribe_.erase(vi_name);
+      ext.subscribeToVehicleInfo(*vi_name);
+      vi_name = vi_waiting_for_subscribe_.erase(vi_name);
```

Comment:
In case of foreach loop you can get rid off saving next to erased item iterator.

Reply:
In this case we need this iterator to guarantee correct order of elements in container during iteration with possible erasing.